korangar/src/graphics/passes/directional_shadow/shader/entity_bindless.wgsl

```diff
@@ -71,20 +73,19 @@ fn vs_main(
     output.original_depth_offset = instance.depth_offset;
     output.original_curvature = instance.curvature;
     output.texture_index = instance.texture_index;
-    output.angle = instance.angle;
+    // can't pass the mat2x2 directly need to pass the instance_index
+    output.instance_index = instance_index;
+    output.alpha = instance.alpha;
     return output;
 }
 
 @fragment
 fn fs_main(input: VertexOutput) -> @builtin(frag_depth) f32 {
-    // Apply the rotation from action
-    let sin_factor = sin(input.angle);
-    let cos_factor = cos(input.angle);
-    let rotate = vec2(input.texture_coordinates.x - 0.5, input.texture_coordinates.y - 0.5) * mat2x2(cos_factor, sin_factor, -sin_factor, cos_factor);
-    let texture_coordinates = vec2(clamp(rotate.x + 0.5, 0.0, 1.0), clamp(rotate.y + 0.5, 0.0, 1.0));
+    let action_rotation_matrix = instance_data[input.instance_index].action_rotation_matrix;
+    let texture_coordinates = action_rotation_transform(input.texture_coordinates, action_rotation_matrix);
```

Comment:
You are using the interpolated value from the fragment shader directly here to apply the action_rotation_transform(). This means you should be able to move this function to the vertex shader. At least in my local test this works perfectly fine!